src/System.Linq/src/System/Linq/Union.cs

```diff
@@ -124,19 +170,101 @@ private Set<TSource> FillSet()
                 return set;
             }
 
-            public TSource[] ToArray()
+            internal override IEnumerable<TSource> GetSourceEnumerable(int index)
             {
-                return FillSet().ToArray();
+                Debug.Assert(index == 0 || index == 1);
+                return index == 0 ? _first : _second;
```

Comment:
As noted elsewhere, I think this should follow the same pattern as in Concat, and return null if the index isn't 0 or 1.